src/bootstrap/src/core/config/tests.rs

```diff
@@ -454,3 +454,53 @@ fn check_rustc_if_unchanged_paths() {
         assert!(config.src.join(p).exists(), "{p} doesn't exist.");
     }
 }
+
+#[test]
+fn override_profile_for_non_git_sources() {
+    let source_path = "path/to/repo.tarball";
+    let is_git_source = source_path.contains(".git");
+
+    if !is_git_source {
+        let mut config = parse(&format!("rust_info.source_path = \"{}\"", source_path));
+
+        if config.profile.is_none() {
+            config.profile = Some("dist".to_string());
+        }
+        assert_eq!(config.profile, Some("dist".to_string()));
+
+        if !is_git_source {
+            assert_eq!(config.profile, Some("dist".to_string()));
+        }
+    }
+}
+
+#[test]
+fn check_if_unchanged_on_non_git_sources() {
+    let source_types = ["tarball", "other-source-type", "path/to/repo.git"];
+
+    for &source_type in source_types.iter() {
+        let is_git_source = source_type.contains(".git");
+
+        if !is_git_source {
+            let if_unchanged_config = parse("llvm.download-ci-llvm = \"if-unchanged\"");
+
+            if !is_git_source {
+                println!(
+                    "ERROR: 'if-unchanged' is not compatible with {} sources. Aborting `download-ci-llvm`.",
+                    source_type
+                );
+                crate::exit!(1);
+            }
+
+            let rustc_if_unchanged_config = parse("rust.download-rustc = \"if-unchanged\"");
+
+            if !is_git_source {
+                println!(
+                    "ERROR: 'if-unchanged' is not compatible with {} sources. Aborting `download-rustc`.",
+                    source_type
+                );
+                crate::exit!(1);
+            }
+        }
+    }
+}
```

Comment:
These tests don't have an enough coverage; they require some more effort to be useful. You can remove them if you want to, so we can handle them in a follow-up PR.